### PR TITLE
Rewrite the lab

### PR DESCRIPTION
**Pull request by The-E**
The main goal of this rewrite is to increase maintainability of the Lab by breaking the monolithic lab.cpp into several smaller subsections.

The basic structure here is as follows:
A central LabManager instance (LMGR) is responsible for opening and closing the Lab screen as well as handling all the onFrame business that needs handling. It keeps track of the global state, and notifies all subdialogs if it changes so that they can update themselves accordingly.

### CODE REVIEW

File: code/lab/labv2.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#include "gamesequence/gamesequence.h"
+
+LabManager* LMGR;
+
+void lab_init() {
+	gr_set_clear_color(0, 0, 0);
+
+	LMGR = new LabManager();
+
+	Game_mode |= GM_LAB;
```
Comment (asarium) [inline code/lab/labv2.cpp:13]: Is this reset anywhere?

**Review by asarium — CHANGES_REQUESTED** [review]
To be honest, this is quite a lot of code to review and I only barely looked at the new lab code. However, since there was a test build with this (at least I thought I saw one on the forums) and these are pretty self-contained it should be ok to merge this and see if any issues arise later.

Marking as "Request changes" since there is one question to clarify but otherwise should be ok.

File: code/lab/labv2.cpp (hunk 1 above)
Comment (The-E) [reply code/lab/labv2.cpp:13]: Yes. When the user leaves the Lab, LabManager::close is called which removes the GM_LAB flag again.

Comment (asarium) [reply code/lab/labv2.cpp:13]: Ah, ok. A bit confusing to do the "set" part here and the "unset" in a different location but nothing critical.

**Review by asarium — APPROVED** [review]

**Pull request merged**